### ISSUE TITLE
RFC: Show publicly published Backstage Plugins on demo.backstage.io

### ISSUE THREAD

**Issue by minkimcello** [backstage/demo#112]
## Motivation

As one of the first initiatives of getting `Backstage for Backstage` started, we want to first generate a list of all publicly published Backstage plugins and add it to the Backstage demo instance.

Although there's already a [plugins page](https://backstage.io/plugins) on the Backstage website, the feature we want to introduce to the demo instance won't be competing or replacing that preexisting plugins page; instead, it'll work hand-in-hand as the new list could be treated as a more comprehensive list. Additionally, the raw list of plugins could even provide a more convenient way of keeping the curated plugins page on the website up to date.

## Approach

1. We'll configure a cron job to search the NPM registry
2. The CI job will need to take the results and ignore packages that aren't Backstage Plugins (see details in the next section); with the filtered list of confirmed plugins we can update the yaml file and create a pull request
3. Once the PR is merged the demo Backstage app will then eventually read the updated list

![overview-diagram](https://raw.githubusercontent.com/minkimcello/backstage.cool/plugins-finder/utils/package_finder/b-horizontal.png)

### Finding the plugins

1. Search NPM for packages with the keyword `backstage`
    - See example payload [here](https://gist.github.com/minkimcello/fa6951765c89bda0e59cf11f409f284a#file-2_npm_search_payload_example-json)
    - See PoC script [here](https://gist.github.com/minkimcello/fa6951765c89bda0e59cf11f409f284a#file-3_npm_search_script-sh)
2. From the search result, we need to filter out the packages that aren't Backstage plugins
    - The [v1 search api](https://github.com/npm/registry/blob/master/docs/REGISTRY-API.md#get-v1search) doesn't provide a way to filter results by package json properties nor does [NPM CLI search](https://docs.npmjs.com/cli/v7/commands/npm-search) which means we need to run `npm view ${pkg} backstage.role` for each package to check if the returned lue is either `frontend-plugin` or `backend-plugin`
3. For the packages that match the criteria, we should append its role to the rest of the metadata:
```diff
[
  {
+    role: "frontend-plugin",
    ...package
  },
  ...
]
```
4. Loop over the list of qualifying packages and create a multi-file yaml file
5. Generate pull request (or push to preexisting PR) to update the list of plugins
    - Once the PR is merged and the updated file is on the primary branch, the ingestion system can pick it up on schedule

### Getting more information

Once we have our list of plugins, we'll want to expand its functionality and offer more information for each plugin. The requirements will vary depending on what information we want to offer to the users.

#### README

There are a few ways to access the README of any package:

1. We can take the package name and version and plug them into `unpkg.com`
2. If we know the exact path of the plugin (for when it's not at the root of the repository), we can use `raw.githubusercontent.com` to retrieve the contents of the readme

#### Code

We may also want to grab the source files in order to add them to the search, but unlike the READMEs there are more hurdles in getting the source files. We obviously can't use `unpkg.com` as it only displays the compiled files of packages; which is fine for markdown or json files but not great for javascript files; so that leaves us with 3 options:

1. The path of the plugins must be specified by developers for the plugins' package.json files
    - Backstage plugins aren't always in its own repository so if we treat the base repository URL as the source of truth, we'll end up ingesting the entire monorepo
2. Clone the entire repository of the provided URLs and search for the package.json that matches the name of the plugin. We can parse its path once we locate the correct package.json
3. Instead of cloning the repository, we can add each repository to the Sourcegraph instance for Backstage and look for the correct package.json via the sourcegraph API
    - This creates unnecessary vendor lock-in for such an essential process, but I wanted to keep this as an option in case option 1 and 2 do not pan out

## TODOs & Questions

- [ ] Complete the script for [`Finding the plugins`](#finding-the-plugins)
  - The script in its [current state](https://gist.github.com/minkimcello/fa6951765c89bda0e59cf11f409f284a#file-3_npm_search_script-sh) only accomplishes step 1
    - I'm considering rewriting the script in golang in order to run the hundreds of executions of `npm view` in parallel
- [ ] Explore what the `repository.directory` property in package.json actually achieves
  - If you take a look at the [package.json of `@backstage/plugin-home`](https://github.com/backstage/backstage/blob/master/plugins/home/package.json#L20), there's a `repository.directory` property set to `"plugins/home"` but that isn't reflected anywhere in the [payload](https://gist.github.com/minkimcello/fa6951765c89bda0e59cf11f409f284a#file-2_npm_search_payload_example-json) from the search of that plugin
    - So either the `directory` property serves another purpose or it wasn't correctly configured for that repository
      - The `links.homepage` value from the payload corresponds to the `homepage` property set at root in the package.json and wasn't generated from the `directory` value

**Comment by taras** [comment]
Awesome work @minkimcello 

> configure a cron job

Do you have any thoughts on the kind of cronjob this would be? Github Actions?

**Comment by minkimcello** [comment]
> Awesome work @minkimcello
> 
> > configure a cron job
> 
> Do you have any thoughts on the kind of cronjob this would be? Github Actions?

Yeah I think it would be best to use Github Actions

**Comment by webark** [comment]
@minkimcello do you think it's worth it to propose a change to the main "plugin creators" https://github.com/backstage/backstage/tree/master/packages/cli/templates to add an additional field to the keywords by default to make this process easier? 

And you are proposing to have the demo site run on an externally managed static yaml file, rather than handle ingestion in some form itself? 🤔 Is this to make it more efficient? Would plugins not have catalog files in them?

**Comment by minkimcello** [comment]
> do you think it's worth it to propose a change to the main "plugin creators" https://github.com/backstage/backstage/tree/master/packages/cli/templates to add an additional field to the keywords by default to make this process easier?

Yeah, we'll eventually need to make a proposal but I was hoping to flesh out our requirements first. In addition to making the search easier, there are also two other things I wanted standardized; 1. the path of the plugin, 2. specifying its corresponding frontend/backend plugin.

But to answer your question, adding a very specific keyword would make the search easier but I think relying on `backstage.role` property will be safer in the long run as it's already mandatory for the plugins to function. A developer might not understand the purpose of a default keyword that was generated in their plugin and there will be a risk of them getting rid of it. What do you think?

> And you are proposing to have the demo site run on an externally managed static yaml file, rather than handle ingestion in some form itself? 🤔 Is this to make it more efficient? Would plugins not have catalog files in them?

Hmmm. You bring up a good point. You're saying instead of running the script via a cron job in CI and feeding the externally processed entities into Backstage, we should consider creating an entity provider instead, right? I think that warrants a conversation with the rest of the group.

**Comment by taras** [comment]
I was also thinking that we'd build an entity provider. @Rugvip how did you see this working?

**Comment by Rugvip** [comment]
@taras yep I think an entity provider would be a good way forward, potentially combining it with a custom location type and processor to scale the processing a bit more and leveraging the processor cache.

@minkimcello regarding the lookups I think it's best to work as close to NPM APIs as possible, rather than shelling out to CLIs that are going to call the API anyway. For grabbing content we should be able to fetch the tarballs straight from NPM too afaik, to for example get hold of the README an additional content. Depending what performance we see in practice it might even be better to just rely on the tarballs for the filtering step, as those contain the current package.json without all of the history.

For example, this gets you the full history and metadata of `@backstage/cli`:

```bash
curl https://registry.npmjs.com/@backstage%2fcli
```

Through that you also get the download URL of all versions and shasums, for example grabbing the latest:

```bash
curl https://registry.npmjs.com/@backstage%2fcli | jq '.versions[."dist-tags".latest].dist.tarball'
```

With that you can then grab the tarball for the package at that version and access files:

```bash
curl https://registry.npmjs.org/@backstage/cli/-/cli-0.22.3.tgz | tar xO package/README.md
```

There might be some rate limiting issues to work around when hitting the API, but we'd likely hit those if we call through the CLI anyway.

**Comment by taras** [comment]
@Rugvip what do you about using an Incremental Entity Provider here? It can automatically take care of backing off when we hit a rate limit. This would keep the processing pipeline clear for processing.

**Comment by Rugvip** [comment]
@taras I'd say we should keep it simple and use a plain entity provider, and then switch over if we hit problems.

**Comment by taras** [comment]
If we're going after simple, then I think we should consider using an incremental entity provider(IEP). An IEP that commits complete entities is much simpler than an entity provider that commits locations that need a smart processor with caching to deal with rate limiting. The downside of using an IEP is that we will loose the ability to trigger refresh. For this reason alone, we should go with a plain entity provider, but not because it's simpler.

For processor caching, do we need a cache backend?

**Comment by Rugvip** [comment]
@taras no need for a cache backend when using the entity cache, that cache is stored in the entity processing table. Regarding IEP I'd love to identify gap in the DX for plain provider + processor that could make it as simple in that case. NPM does have webhooks so it could be good to build for that too. No super strong opinions either but I think plain *should* be a better fit since the data volume is low.

**Comment by Rugvip** [comment]
@minkimcello I wanted to double check alignment on the scope of the plugin browsing on the demo site, since there's been some back and forth around this in the discussions in the adopter SIG.

The piece I want to make sure we're aligned on is that there is a clear distinction between the purpose of the plugin browsing on the demo site vs the microsite (backstage.io). The plugin browsing experience on the demo site would be tailored towards Backstage plugin builders. For example, we'd have technical documentation for how to extend the plugin, perhaps build status, versions, and so on.

On the other hand, the plugin browsing on the microsite would be tailored for Backstage adopters/integrators. We would also avoid mixing the two, and in particular I think we want to avoid a situation where Backstage adopters need to go to the demo site to browse plugins, and instead keep that contained in the microsite. The main piece that I'm interested in there is to try to keep the adoption experience as simple as possible, with all browsing on backstage.io. We'll of course link out to different places, but avoid the need to bounce between different sites.

What that means in practice is that we would keep the plugin browsing experience on the demo site focused around developing plugins and avoid features for adopters. Examples of adopter-centric features would be for example plugin scoring and categorization, extensive user docs, or installation tools. Those are all things I think it's better if we add only to the microsite instead.

Implementation wise I think we can mix things up a bit more though, for example relying on the catalog data on the demo site to populate the microsite.

I had a chat around this with @taras as well yesterday. Let me know if you have any thoughts or don't agree with this direction. If it's all good, would you mind adding a small piece to the RFC around this distinction? Perhaps under motivation or in a separate section. And feel free to steal text :grin:

**Comment by webark** [comment]
> What that means in practice is that we would keep the plugin browsing experience on the demo site focused around developing plugins and avoid features for adopters.

One of the main places that this initiative started from was that people looking to adopt backstage don't have a clear understanding for what it does, what it looks like, and what its capabilities are. And having a more fleshed out and real world example would help a lot of people in that adoption phase. So why would we want to avoid features for some of the users and use cases we are wanting to target?

**Comment by taras** [comment]
@webark the risk is that we'll invest a lot of effort into the Backstage for Backstage(B4B) and the microsite will miss out on improvement that'll be helpful to everyone, not just users of B4B. We can use this as a rule of thumb as we get started. In the future, we could revisit this rule. 

I'm kinda wondering if we're experiencing what some of the adopters have been asking for. Adopters need a way to create a microsite for subset of content in their Backstage portal. Maybe out of this work a solution will emerge that we can use for the microsite and adopters who need a public microsite.

**Comment by webark** [comment]
Maybe. And by updates that the microsite would miss out on, are we talking about the backstage core docs? Or are there some other items planned that are more focused on showcasing and promoting its feature set other than what's on the homepage and the "what is backstage" part of the docs (which should probably be moved into its own section now that I think of it)

**Comment by minkimcello** [comment]
@Rugvip I'm all for making a clear distinction between the two (microsite and catalog) and I was trying to find a way to express that better in the RFC. I agree that the two should serve separate purposes so it's not just a mishmash of information scattered everywhere.

However, I thought the distinction we were trying to make was that the microsite would display a curated list of "approved" plugins (as not to overwhelm anyone) where adopters can get a good idea of what they could use for their Backstage instance. An adopter might not find a specific plugin they had in mind and the microsite could go "these are just a few plugins, for the full list and more information on each plugin, go to the catalog".

Also, I'm having trouble drawing the line between adopters and integrators. I assumed it was a bit of a spectrum. Don't adopters eventually become integrators? 😅

---

> the risk is that we'll invest a lot of effort into the Backstage for Backstage(B4B) and the microsite will miss out on improvement that'll be helpful to everyone. We can use this as a rule of thumb as we get started. In the future, we can revisit this rule.

@taraas I think once the microsite's plugin page starts utilizing the catalog to populate its list, wouldn't it be tightly coupled enough that we don't need to worry about this as much?

**Comment by taras** [comment]
> @taras I think once the microsite's plugin page starts utilizing the catalog to populate its list, wouldn't it be tightly coupled enough that we don't need to worry about this as much?

I think so

**Comment by minkimcello** [comment]
My concern is that we might need to hold back from adding cool things to the demo for the sake of the microsite. There will definitely be an overlap of the demo and the microsite as there is no way of drawing a clear line of what's only useful to developers and what's only useful to adopters.

@Rugvip so instead of trying to make sure the demo doesn't include anything that adopters will find appealing, can we try to improve both the microsite and the demo in parallel? That way, for each feature, we only need to ask "is this valuable to developers" and not "is this valuable to developers and irrelevant to adopters?"

**Comment by Rugvip** [comment]
@minkimcello yep no strong feelings on them keeping parallel, my concern is with the demo being ahead

**Comment by awanlin** [comment]
Hi @minkimcello, just wanted to know if anything further has come of this or if we should close this RFC?
